pypicloud/access/_ldap.py

```diff
@@ -0,0 +1,391 @@
+        try:
+            admins = res[0][1][LDAP._admin_field]
+        except (IndexError, KeyError) as err:
+            logging.warn("Error retrieving admins from %s: %r", admin_dn, err)
+            return
+
+        for admin in admins:
+            try:
+                LDAP._admin_usernames.append(
+                    LDAP._server.search_s(
+                        admin,
+                        ldap.SCOPE_SUBTREE,
+                    )[0][1][LDAP._id_field][0]
+                )
+            except (IndexError, KeyError) as error:
+                logging.warn("Error looking up admin %s: %r", admin, error)
+            else:
+                LDAP._admins.append(admin)
+
+    @staticmethod
+    def admins():
+        """Returns a list of all the admin DNs."""
+
+        if not hasattr(LDAP, "_admins"):
+            LDAP._admins = []
+            LDAP._admin_usernames = []
+            for admin_dn in LDAP._admin_dns:
+                LDAP._add_admins_from_dn(admin_dn)
+
+        return LDAP._admins
+
+    @staticmethod
+    def admin_usernames():
+        """Returns a list of the admin usernames."""
+
+        if not hasattr(LDAP, "_admin_usernames"):
+            LDAP.admins()
+        return LDAP._admin_usernames
+
+    @staticmethod
+    def user_dn(username):
+        """Returns the dn for the username."""
+
+        return LDAP._all_users.get(username)
+
+    @staticmethod
+    @reconnect
+    def bind_user(user_dn, password):
+        """Attempts to bind as the user, then rebinds as service user again."""
+
+        try:
+            LDAP._server.simple_bind_s(user_dn, password)
+        except ldap.INVALID_CREDENTIALS:
+            return False
+        else:
+            return True
+        finally:
+            LDAP._connect()
+
+
+class LDAPAccessBackend(IAccessBackend):
+    """
+    This backend allows you to store all user and package permissions in a SQL
+    database and authenticate against a LDAP server
+    """
+
+    DEFAULT_PERMISSIONS = {
+        "admin": ("read", "write"),
+        "authenticated": ("read", "write"),
+        "everyone": ("read",),
+    }
+
+    def __init__(self, request=None, **kwargs):
+        self.request = request
+        super(LDAPAccessBackend, self).__init__(request, **kwargs)
+
+    @classmethod
+    def configure(cls, settings):
+        kwargs = super(LDAPAccessBackend, cls).configure(settings)
+        LDAP.configure(settings)
+        return kwargs
+
+    def allow_register(self):
+        return False
+
+    def _get_password_hash(self, *_):
+        return ""
+
+    def verify_user(self, username, password):
+        """
+        Check the login credentials of a user
+
+        For Mutable backends, pending users should fail to verify
+
+        Parameters
+        ----------
+        username : str
+        password : str
+
+        Returns
+        -------
+        valid : bool
+            True if user credentials are valid, false otherwise
+
+        """
```

Comment:
I think you can exclude most of these lengthy docstrings if you want, since they're on the base class. Up to you.